### ISSUE TITLE
[Feature] 💥 Upgrade to webpack 4 🎉👍

### ISSUE THREAD

**Issue by almothafar** [gdi2290/angular-starter#1973]
* **I'm submitting a ...**
[ ] bug report
[X] feature request
[ ] question about the decisions made in the repository

* **Do you want to request a *feature* or report a *bug*?**
Feature: Upgrade to webpack 4

* **What is the current behavior?**
Currently, it is Webpack 3

* **What is the expected behavior?**
Use Webpack 4 as it contains big improvements

* **What is the motivation / use case for changing the behavior?**
As you can see the big features and changes in release notes: https://github.com/webpack/webpack/releases/tag/v4.0.0

🎉 💃

**Comment by lanovoy** [comment]
Looks like will have to wait until html-webpack-plugin is upgraded. Might have to upgrade some others as well, but this one is crucial.

**Comment by Uter1007** [comment]
https://github.com/jantimon/html-webpack-plugin/issues/860

Looks like they are already working on it. Maybe you can test further with the alternative: npm install -D webpack-contrib/html-webpack-plugin

**Comment by lanovoy** [comment]
It fails to install for me. Runs some specs and throws lots of exceptions. May be because we're using corporate npm registry and it doesn't have an access to webpack-contrib/ .

**Comment by sherlock1982** [comment]
There are a couple of packages which needs patches. https://github.com/szrenwei/inline-manifest-webpack-plugin/issues/10 and https://github.com/numical/script-ext-html-webpack-plugin/issues/41. Patches are obvious and I used patch-package for now.

**Comment by lanovoy** [comment]
FYI, Angular CLI has been updated for webpack 4.0

**Comment by almothafar** [comment]
@sherlock1982 we can use `webpack-inline-manifest-plugin` instead of `inline-manifest-webpack-plugin`
And `script-ext-html-webpack-plugin` version 2 is supporting webpack 4

**Comment by lanovoy** [comment]
I've did a quick and dirty upgrade to the latest version of webpack (4.1.1). It's working, the build is much faster, but unfortunately I had to disable html-elements-plugin which exists as part of this repo (angular-starter\config\html-elements-plugin\index.js) since it's not compatible and I'm lacking both knowledge and time to upgrade it. My code changes are uploaded into the branch [>> here <<](https://github.com/gdi2290/angular-starter/tree/feature/migrate-to-webpack4). For now I'm not going to create PR for it. If someone is willing to look into it, please feel free.

**Comment by almothafar** [comment]
@lanovoy why you removed the `CommonsChunkPlugin`? 
https://github.com/gdi2290/angular-starter/compare/feature/migrate-to-webpack4#diff-f737fedb5357468044db2dbe98eb0218L189

Is there any reason? I think chunks are helpful, isn't?

for `html-elements-plugin` I think it can be a PR for it, sounds the Author is active so we can get this running once it is done, and maybe some people using it.

I'm not sure why you removed the `ModuleConcatenationPlugin` as well, I see commit says deprecated plugin, which one is? because I don't see the source of this info in their doc, is it NamedModulesPlugin or ModuleConcatenationPlugin.

There are my notes currently.

**Comment by kuhnroyal** [comment]
https://medium.com/webpack/webpack-4-code-splitting-chunk-graph-and-the-splitchunks-optimization-be739a861366

**Comment by almothafar** [comment]
@kuhnroyal thanks, but at least add some context to your link, like a TL;DR version.

**Comment by lanovoy** [comment]
Yes, those plug-ins were removed as they are listed as not needed for
webpack anymore. Webpack does the same as they do out of the box now.

вс, 18 марта 2018 г., 15:39 Al-Mothafar Al-Hasan <notifications@github.com>:

**Comment by lanovoy** [comment]
ngtools 1.10.0 (and 1.10.2) are technically not compatible with the Webpack
4, but you may be right and it's just a cosmetic incompatibility... Feel
free to verify yourself.

вс, 18 марта 2018 г., 15:54 Maksym Lanovyi <lanovoy@gmail.com>:

**Comment by almothafar** [comment]
@lanovoy I'll do verify later for sure, I just did code review and was looking for clarifications before I reinvent the wheel, sadly not have enough time to do that by myself, so I appreciate your efforts.
If I have fixes I'll do that on the branch that you created.

**Comment by lanovoy** [comment]
Awesome! Thank you

вс, 18 марта 2018 г., 16:58 Al-Mothafar Al-Hasan <notifications@github.com>:

**Comment by lanovoy** [comment]
One another outstanding issue with the webpack 4 conversion in the branch created by me, is that there is some dependency to webpack 3.x still somewhere (likely in the plugins npm packages). After npm install you will find both webpack versions in the node_modules.

**Comment by govi2010** [comment]
@lanovoy let me look into it. I am not expert but I can try.

**Comment by govi2010** [comment]
@lanovoy  I checked your code. use `"@angular/cli": "next",` instead of `"@angular/cli": "^1.7.3",` becasue `"@angular/cli": "^1.7.3",` version has dependency of webpack 3

**Comment by almothafar** [comment]
I was thinking about postponing this, until Angular 6 is released it should be on 4th of April 
https://github.com/angular/angular/blob/master/docs/RELEASE_SCHEDULE.md

**Comment by lanovoy** [comment]
I agree, it would be the best to wait until Angular 6 release. I will
create an issue to upgrade shipped with this repo plug-in, tho...

пн, 19 марта 2018 г., 15:38 Al-Mothafar Al-Hasan <notifications@github.com>:

**Comment by lanovoy** [comment]
#1995

Additionally in my branch there is a problem with the karma test runner it seems so.

**Comment by govi2010** [comment]
I agree with waiting for angular 6 release decision. I will try to update html-elements-plugin I don't know how to update but will give a try.

**Comment by Yizhachok** [comment]
@lanovoy have you checked HMR reload after code updating? Because after CommonsChunkPlugin removing it was broken to me.

**Comment by sherlock1982** [comment]
@lanovoy you're using jasmine-core 3.1.0 which has an issue with current zone.js. This is explained [here](https://github.com/angular/zone.js/issues/1015). You could use 2.9.1 or wait until new zone.js release.

**Comment by lanovoy** [comment]
@Yizhachok Nope, didn't test HMR. You can clone from my branch and verify yourself, tho. 
In any case HMR is not really working with current setup (pre-upgrade to webpack4) - it's doing whole page reload instead of hot swapping changed parts...

**Comment by lanovoy** [comment]
@sherlock1982 Thanks. I suppose we're going to wait for Angular 6 release in any case, so hopefully jasmine-core will be updated by that time as well...

**Comment by mrowles** [comment]
Heads up, there are few issues with current starter test webpack + ts config using packages webpack@4 + typescript @2.6.2 + awesome-typescript-loader@5. More info [here](https://github.com/s-panferov/awesome-typescript-loader/issues/534)

**Comment by lanovoy** [comment]
Looks like Angular 6 release is postponed again. RC3 just got released. Many really sweet features with Angular 6, so upgrading to it should be pretty much a priority for this starter project.

**Comment by konuch** [comment]
Bump, **Angular 6.0** [**is out**](https://github.com/angular/angular/releases/tag/6.0.0).

**Comment by govi2010** [comment]
@konuch  We were waiting for angular 6 release. Yes we need to test our `migrate-to-webpack4` branch before we push to master. 
I will start working on SSR implementation this week end.

**Comment by aaronwJordan** [comment]
@govi2010 I did a quick skim in `migrate-to-webpack4` and didn't see anything for SSR. Is that coming soon? We're looking at SSR soon and I want to do it the 'correct` way. Thanks!

**Comment by lanovoy** [comment]
It was merged into master. Thanks for the help people! PR for reference https://github.com/gdi2290/angular-starter/pull/2019
